### DISCUSSION TITLE
Type checking for mapped column declarations

### DISCUSSION THREAD

**Discussion by acanizares** [sqlalchemy/sqlalchemy#11093] [Ideas]
SQLAlchemy v2 already brings huge typing improvements. Thanks for doing this!

The bit that I'm missing is type checking in the mapped columns declaration: 

1. `mapped_column` returns `MappedColumn[Any]`, so there is no guarantee that the type annotation matches the SQLA type engine passed to `mapped_column` (if any). I know that this usage is not the expected one. You should use either the type annotation or the type engine argument. But if you accidentally use both, then the type checker will validate your code according to the type annotation, while SQLA will use the `TypeEngine` type at runtime, which will result in runtime errors if both types are incompatible.

This code should result in a type error:

```python
class Foo(Base):
    __tablename__ = "foo"
    # This should be a type error
    x: Mapped[int] = mapped_column(sqlalchemy.String)
```

2. When using `MappedAsDataclass`, it would be even nicer if we could check the type of the default value passed to `mapped_column`: the type of `default` should match the annotated type. In particular, `default=None` should only be possible for columns annotated with an optional type.

```python
class Foo(MappedAsDataclass, Base):
    __tablename__ = "foo"
    # This should be a type error
    x: Mapped[int] = mapped_column(default=None)
    # This should be ok
    y: Mapped[int | None] = mapped_column(default=None)
    # This should be ok too
    x: Mapped[int] = mapped_column(default=7)
```
Side note: for `relationships` we probably want to allow passing `default=None` for non nullable columns, since such columns might be back populated after the object has first been initialized?

The following changes to [mapped_column](https://github.com/sqlalchemy/sqlalchemy/blob/b5e8c80f24c326e8491d1db6c5b32dcf730b53b4/lib/sqlalchemy/orm/_orm_constructors.py#L98) worked locally for me as desired (with both Mypy and Pyright). Does this sound like a reasonable approach? If so then I can make a pull request.
I'm omitting lines with no changes. The overloaded variant would only be necessary to check the type of the default value.

```python
@overload
def mapped_column(
    ...
    default: Union[_NoArg, _T] = _NoArg.NO_ARG,
   ...
) -> MappedColumn[_T]:
    ...

def mapped_column(
    __name_pos: Optional[
        Union[str, _TypeEngineArgument[_T], SchemaEventTarget]
    ] = None,
    __type_pos: Optional[
        Union[_TypeEngineArgument[_T], SchemaEventTarget]
    ] = None,
    ...
    default: Optional[Union[_NoArg, _T]] = _NoArg.NO_ARG,
    ...
) -> MappedColumn[Optional[_T]]:
```

**Comment by CaselIT** [comment]
Hi,

Regarding 1 I think it was one of the objective to allow overriding the type separately from the python type, but I guess it does not make much sense to type something as int and the have String as the type, so I would think it may still be ok.
Regarding 2 it's probably reasonable

Thoughs on this @zzzeek ?

**Reply by acanizares** [reply]
@CaselIT Thanks for your reply! I worked a bit on this and came to the conclusion that it is simply not possible to express the type of `mapped_column` correctly as it is implemented right now. I believe a prerequisite would be to change the default value of `nullable` to `False`. 

Because otherwise there is no way to type something like `mapped_column(Integer)` correctly. `mapped_column(Integer)` can either be of type `MappedColumn[int]` or `MappedColumn[Optional[int]]`, and the default should be the more specific type of those two, namely `MappedColumn[int]` since both of the following assignments should be valid:

```
x: Mapped[int]  = mapped_column(Integer)
y: Mapped[Optional[int]]  = mapped_column(Integer)
```

I created a [pull request](https://github.com/sqlalchemy/sqlalchemy/pull/11154) to share with you the typing changes. The [tests](https://github.com/sqlalchemy/sqlalchemy/blob/99cac9a3a5a767cb5ac2da9a10b952b2ac74ce7d/test/typing/plain_files/orm/mapped_column.py#L95) give a good overview of the typing improvements. 
It would still be necessary to change the default behaviour for types to be correct though. Even though it's extremely unlikely that you make this change I wanted to share this work with you :)

**Reply by CaselIT** [reply]
I guess we can't make the default just be `MappedColumn[Optional[int]]` since type checkers would complain when assigned to `Mapped[int]` regardless of the variance of the generic, correct?

**Reply by acanizares** [reply]
@CaselIT correct. `Mapped` is covariant and I believe we'd need it to be contravariant as well, which is not possible.

**Reply by CaselIT** [reply]
We can't do anything using the fact that the default for nullable is `NULL_UNSPECIFIED`?

meaning that the type checkers would complain only if doing
```py
foo: Mapped[int] = mapped_column(nullable=True)
```
I guess this the opposite is not possible to flag, since it's not an error to do `value: int|None=5`
```py
# this can't be flagged
bar: Mapped[int|None] = mapped_column(nullable=False)
```

The default case would be ok in both cases:
```py
foo: Mapped[int] = mapped_column()
bar: Mapped[int|None] = mapped_column()
```

**Reply by acanizares** [reply]
But what should be the type of, say, `mapped_column(Integer)`? In order for both `foo` and `bar` in your example to be ok for the typechecker we need to define it as `MappedColumn[int]`. But that is a lie :). This is problematic when we have no type annotation
```
x = mapped_column(Integer)
```
because `x` would be typed as `int`, but this is wrong because it could also be `None`, right?

Not sure this answers your question or just poses a new one 😅

**Reply by CaselIT** [reply]
You have convinced me that what sqlalchemy does is not possible to express using typing, like you mentioned about. Thanks!

**Comment by zzzeek** [comment]
hi -

from looking at some of the changes in the PR I have to clarify SQLAlchemy.    SQLAlchemy ORMs job is to mediate the line between a Python object model behavior and the nuts-and-bolts details of a database schema.    Real world scenarios will involve that database columns might have lots of unusual things going on that dont indicate how the typing will be expressed in the object model, such as nullability or SQL-level defaults.

example one, type.   This is arguably the most appropriate for a type error.  if someone is trying to use some custom type that deals with strings being coerced to ints, it would be a typedecorator that can declare the correct type, OK.

```
class Foo(Base):
    __tablename__ = "foo"
    x: Mapped[int] = mapped_column(sqlalchemy.String)
```

example two

```
class Foo(MappedAsDataclass, Base):
    __tablename__ = "foo"
    # This should be a type error
    x: Mapped[int] = mapped_column(default=None)
```

so when using mappedasdataclass, "default" takes on the meaning it has in a dataclass field.  in that case, we follow whatever dataclasses does, so this one is valid.  However if we are talking about `insert_default`, which is the SQL level default, that can absolutely be `None` and should be independent of object model typing.

in the PR though I saw a lot of this one:

```
class Foo(Base):
    __tablename__ = "foo"
    x: Mapped[int] = mapped_column(nullable=True)
```

that above is valid.  This class might be mapped to a database table that for legacy or backend-specific reasons, a particular column has to support nullability, even though in practice within the model, the attribute should never be None.    that's documented here: https://docs.sqlalchemy.org/en/20/orm/declarative_tables.html#mapped-column-derives-the-datatype-and-nullability-from-the-mapped-annotation

specific example, MySQL GENERATED AS.  The columns "area" and "perimeter" below will **always** have a value when the object is retrieved, the database generates them:

```py
from __future__ import annotations

from sqlalchemy import Computed
from sqlalchemy import create_engine
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column


class Base(DeclarativeBase):
    pass

class Square(Base):
    __tablename__ = "square"
    id: Mapped[int] = mapped_column(primary_key=True)
    side: Mapped[int]
    area: Mapped[int] = mapped_column(Computed("side * side"), nullable=False)
    perimeter: Mapped[int] = mapped_column(Computed("4 * side"), nullable=False)

e = create_engine("mysql://scott:tiger@localhost/test")
Base.metadata.create_all(e)

```

however, this program fails:

```
sqlalchemy.exc.ProgrammingError: (MySQLdb.ProgrammingError) (1064, "You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near 'NOT NULL, \n\tperimeter INTEGER GENERATED ALWAYS AS (4 * side) NOT NULL, \n\tPRIM...' at line 4")
```

MySQL requires computed columns to be NULL even though they will never be NULL, so this program **is** valid:

```
class Square(Base):
    __tablename__ = "square"
    id: Mapped[int] = mapped_column(primary_key=True)
    side: Mapped[int]
    area: Mapped[int] = mapped_column(Computed("side * side"), nullable=True)
    perimeter: Mapped[int] = mapped_column(Computed("4 * side"), nullable=True)

```

now to be fair, our MySQL dialect automatically does the correct NULL if you omit nullable entirely.    But that's a case we cover.  we have no idea what other cases might exist where people have triggers or unusual backends where they want particular schema directives to be present for whatever reason.

What should be noted here is that the ORM annotated declarative system is attempting to automate the process of designing a schema using Python type annotations.  However, **it is not intended to be bidirectional**.  The specifics of a database schema, which can be affected by all kinds of driver / backend / legacy issues should **always be completely independent of the object model** - that is the point of the object relational mapper, which is to **mediate** translation of data from a DB schema and an object model.

So even though we probably could handle `Mapped[int] = mapped_column(String)` being a type error, in the bigger sense trying to have the typing rules flow backwards from schema detail to python type declarations is not in the spirit of what we have here.   someone saying `Mapped[int] = mapped_column(String)` is trying to do something very specific; if they want the `Mapped[int]` to be mapped to `Integer` they'd leave the type blank on `mapped_column()`.  this example is stretching it, but in the bigger sense, the details sent to `mapped_column()` are implementation details of a database schema and are expected to be independent of object model behavior.

**Reply by acanizares** [reply]
Oh wow, thanks. I really appreciate your detailed answer. The MySQL example makes a great point. 
In my company's codebase mismatches between the type annotation and the `mapped_column` parameters will almost always be by mistake, so we'll need to find an alternative way of preventing these. Such mistakes are also probably easy to make when migrating a large codebase to SQLA 2.

Anyways, thanks for your insights! I'll close the PR and the discussion.

**Reply by CaselIT** [reply]
This seems like something that could be developed on top of mapped_colum that's more pedantic regarding typing.
Like a function that has more type validation than mapped_column but that in the end calls it. This could maybe be provided as an example.

But I'm not sure if we the nullable issue can be overcome in any way (other tham making nullable required or reversing the default only in this hypothetical function)
